File: app/src/main/java/com/example/punchpad2/FolderAdapter.java
package com.example.punchpad2;

import android.content.Context;
import android.view.GestureDetector;
import android.view.LayoutInflater;
import android.view.MotionEvent;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageButton;
import android.widget.LinearLayout;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

public class FolderAdapter extends RecyclerView.Adapter<FolderAdapter.FolderViewHolder> {

    private final List<FolderWithNotes> folders;
    private final Context context;
    private final OnNoteInteractionListener listener;

    private boolean deleteMode = false;
    private boolean foldersOnlyView = false;
    private final Set<NoteEntity> expandedNotes = new HashSet<>();
    public int getAdapterPositionForNote(NoteEntity target) {
        for (int i = 0; i < folders.size(); i++) {
            for (NoteEntity note : folders.get(i).notes) {
                if (note.id == target.id) return i;
            }
        }
        return RecyclerView.NO_POSITION;
    }

    public interface OnNoteInteractionListener {
        void onNoteTapped(NoteEntity note);
    }

    public FolderAdapter(Context context, List<FolderWithNotes> folders, OnNoteInteractionListener listener) {
        this.context = context;
        this.folders = folders;
        this.listener = listener;
    }

    public void setDeleteMode(boolean deleteMode) {
        this.deleteMode = deleteMode;
        notifyDataSetChanged();
    }

    public void setFoldersOnlyView(boolean foldersOnlyView) {
        this.foldersOnlyView = foldersOnlyView;
        notifyDataSetChanged();
    }

    public void updateFilteredList(List<FolderWithNotes> filteredFolders, boolean foldersOnly) {
        this.folders.clear();
        this.folders.addAll(filteredFolders);
        this.foldersOnlyView = foldersOnly;
        notifyDataSetChanged();
    }

    @NonNull
    @Override
    public FolderViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View v = LayoutInflater.from(parent.getContext()).inflate(R.layout.item_folder, parent, false);
        return new FolderViewHolder(v);
    }

    @Override
    public void onBindViewHolder(@NonNull FolderViewHolder holder, int position) {
        holder.bind(folders.get(position));
    }

    @Override
    public int getItemCount() {
        return folders.size();
    }

    class FolderViewHolder extends RecyclerView.ViewHolder {
        TextView folderName;
        LinearLayout notesContainer;
        ImageButton expandButton;

        FolderViewHolder(View itemView) {
            super(itemView);
            folderName = itemView.findViewById(R.id.folderName);
            notesContainer = itemView.findViewById(R.id.notesContainer);
            expandButton = itemView.findViewById(R.id.expandButton);
        }

        void bind(FolderWithNotes folderWithNotes) {
            folderName.setText(folderWithNotes.folder.name);
            notesContainer.removeAllViews();

            expandButton.setOnClickListener(v -> {
                folderWithNotes.folder.expanded = !folderWithNotes.folder.expanded;
                notifyDataSetChanged();
            });

            expandButton.setRotation(folderWithNotes.folder.expanded ? 180 : 0);
            notesContainer.setVisibility(folderWithNotes.folder.expanded && !foldersOnlyView ? View.VISIBLE : View.GONE);

            if (folderWithNotes.folder.expanded && !foldersOnlyView) {
                List<NoteEntity> notes = folderWithNotes.notes;
                NoteAdapter tempAdapter = new NoteAdapter(context, notes, deleteMode, null);

                for (int i = 0; i < Math.min(3, notes.size()); i++) {
                    View noteView = LayoutInflater.from(context).inflate(R.layout.item_note, notesContainer, false);
                    NoteAdapter.NoteViewHolder noteHolder = tempAdapter.new NoteViewHolder(noteView);
                    NoteEntity note = notes.get(i);
                    noteHolder.bind(note);

                    setupNoteClickListeners(noteView, note);
                    notesContainer.addView(noteView);
                }
            }
        }

        private void setupNoteClickListeners(View noteView, NoteEntity note) {
            GestureDetector gestureDetector = new GestureDetector(context, new GestureDetector.SimpleOnGestureListener() {
                @Override
                public boolean onSingleTapConfirmed(MotionEvent e) {
                    listener.onNoteTapped(note);
                    return true;
                }

                @Override
                public boolean onDoubleTap(MotionEvent e) {
                    // double-tap behavior removed—favorites handled elsewhere
                    return true;
                }

                @Override
                public void onLongPress(MotionEvent e) {
                    // reserved for future use
                }
            });

            noteView.setOnTouchListener((v, event) -> gestureDetector.onTouchEvent(event));
        }
    }

    private void handleNoteExpansion(NoteEntity note) {
        if (!expandedNotes.contains(note)) {
            if (expandedNotes.size() >= 3) {
                NoteEntity toCollapse = new ArrayList<>(expandedNotes).get(0);
                toCollapse.expanded = false;
                expandedNotes.remove(toCollapse);
            }
            note.expanded = true;
            expandedNotes.add(note);
        } else {
            note.expanded = false;
            expandedNotes.remove(note);
        }
    }
}

File: app/src/main/java/com/stratonotes/FolderAdapter.kt
package com.example.punchpad2

import android.content.Context
import android.graphics.drawable.Drawable
import android.graphics.drawable.GradientDrawable
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.view.inputmethod.EditorInfo
import android.widget.*
import androidx.core.content.ContextCompat
import androidx.core.graphics.ColorUtils
import androidx.core.graphics.drawable.DrawableCompat
import androidx.recyclerview.widget.RecyclerView
import com.stratonotes.FolderWithNotes
import com.stratonotes.NoteEntity
import com.stratonotes.UserColorManager

class FolderAdapter(
    private val context: Context,
    private val folders: MutableList<FolderWithNotes>,
    private val listener: (NoteEntity, Boolean) -> Unit,
    private val noteLayoutResId: Int
) : RecyclerView.Adapter<FolderAdapter.FolderViewHolder>() {

    fun getFolderIndexById(folderId: Long): Int {
        return folders.indexOfFirst { it.folder.id == folderId }
    }

    private val folderStates = mutableMapOf<Long, ExpandMode>()
    private val folderLoadedCounts = mutableMapOf<Long, Int>()
    private var selectionMode = false
    private val selectedNotes = mutableSetOf<NoteEntity>()

    enum class ExpandMode {
        PARTIAL, FULL, COLLAPSED
    }

    fun getSelectedNotes(): List<NoteEntity> = selectedNotes.toList()

    fun exitSelectionMode() {
        selectionMode = false
        selectedNotes.clear()
        notifyDataSetChanged()
    }

    fun updateFilteredList(filteredFolders: List<FolderWithNotes>) {
        folders.clear()
        folders.addAll(filteredFolders)
        filteredFolders.forEach { folder ->
            val id = folder.folder.id
            if (!folderStates.containsKey(id)) {
                folderStates[id] = ExpandMode.PARTIAL
                folderLoadedCounts[id] = 50
            }
        }
        selectedNotes.clear()


        notifyDataSetChanged()
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): FolderViewHolder {
        val v = LayoutInflater.from(parent.context).inflate(R.layout.item_folder, parent, false)
        return FolderViewHolder(v)
    }

    override fun onBindViewHolder(holder: FolderViewHolder, position: Int) {
        val folder = folders[position]

        val context = holder.itemView.context
        val folderColor = UserColorManager.getFolderColor(context)

        val tabView = holder.itemView.findViewById<View>(R.id.folderHeader)
        val drawable = ContextCompat.getDrawable(tabView.context, R.drawable.folder_header_background)?.mutate()
        (drawable as? GradientDrawable)?.setColor(folderColor)
        tabView.background = drawable

        holder.folderName.setBackgroundColor(folderColor)

        holder.bind(folder)
    }

    override fun getItemCount(): Int = folders.size

    inner class FolderViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        val folderName: TextView = itemView.findViewById(R.id.folderName)
        val notesContainer: LinearLayout = itemView.findViewById(R.id.notesContainer)

        fun bind(folderWithNotes: FolderWithNotes) {
            val folderId = folderWithNotes.folder.id
            val folderText = folderWithNotes.folder.name ?: "(Unnamed)"
            folderName.text = folderText

            val expandMode = folderStates.getOrDefault(folderId, ExpandMode.PARTIAL)
            val maxCount = folderLoadedCounts.getOrDefault(folderId, 50)

            notesContainer.removeAllViews()
            notesContainer.visibility =
                if (expandMode == ExpandMode.COLLAPSED) View.GONE else View.VISIBLE

            val notesToShow = when (expandMode) {
                ExpandMode.COLLAPSED -> emptyList()
                ExpandMode.PARTIAL -> folderWithNotes.notes.take(3)
                ExpandMode.FULL -> folderWithNotes.notes.take(maxCount)
            }

            val noteColor = UserColorManager.getNoteColor(context)
            val folderColor = UserColorManager.getFolderColor(context)

            notesToShow.forEach { note ->
                val noteView = LayoutInflater.from(context).inflate(noteLayoutResId, notesContainer, false)
                val noteText = noteView.findViewById<TextView>(R.id.noteText)
                val starIcon = noteView.findViewById<ImageView>(R.id.starIcon)
                val checkbox = noteView.findViewById<CheckBox?>(R.id.noteCheckbox)
                val fadeView = noteView.findViewById<View>(R.id.noteFade)

                val bgRes = if (note.content.length > 150) {
                    R.drawable.note_expanded_background
                } else {
                    R.drawable.note_preview_background
                }

                ContextCompat.getDrawable(context, bgRes)?.mutate()?.let { drawable ->
                    DrawableCompat.setTint(drawable, noteColor)
                    noteView.background = drawable
                } ?: noteView.setBackgroundColor(noteColor)

                if (note.content.length > 150) {
                    val fadeBottom = folderColor
                    val fadeTop = ColorUtils.setAlphaComponent(folderColor, 0)
                    val gradient = GradientDrawable(
                        GradientDrawable.Orientation.BOTTOM_TOP,
                        intArrayOf(fadeBottom, fadeTop)
                    )
                    gradient.cornerRadius = 0f
                    fadeView.background = gradient
                    fadeView.visibility = View.VISIBLE
                } else {
                    fadeView.visibility = View.GONE
                }

                starIcon.setImageResource(if (note.isFavorite) R.drawable.ic_star_filled else R.drawable.ic_star_outline)
                starIcon.visibility = View.VISIBLE

                noteText.text = note.content

                noteText.setOnLongClickListener {
                    if (!selectionMode) {
                        selectionMode = true
                        selectedNotes.clear()
                    }
                    selectedNotes.add(note)
                    notifyDataSetChanged()
                    true
                }

                noteText.setOnClickListener {
                    if (selectionMode) {
                        val isChecked = !selectedNotes.contains(note)
                        if (isChecked) selectedNotes.add(note) else selectedNotes.remove(note)
                        notifyDataSetChanged()
                    } else {
                        listener(note, false)
                    }
                }


                if (selectionMode) {
                    checkbox?.visibility = View.VISIBLE
                    checkbox?.isChecked = selectedNotes.contains(note)
                } else {
                    checkbox?.visibility = View.GONE
                }

                checkbox?.setOnClickListener {
                    val isChecked = checkbox.isChecked
                    if (isChecked) selectedNotes.add(note) else selectedNotes.remove(note)
                }

                starIcon.setOnClickListener {
                    note.isFavorite = !note.isFavorite
                    starIcon.setImageResource(if (note.isFavorite) R.drawable.ic_star_filled else R.drawable.ic_star_outline)

                }

                notesContainer.addView(noteView)
            }

            val totalNotes = folderWithNotes.notes.size
            if (expandMode == ExpandMode.FULL && totalNotes > notesToShow.size) {
                val loadMore = TextView(context).apply {
                    text = "Load More"
                    setPadding(16, 16, 16, 16)
                    setTextColor(context.getColor(R.color.white))
                    setOnClickListener {
                        folderLoadedCounts[folderId] = folderLoadedCounts.getOrDefault(folderId, 50) + 50
                        notifyItemChanged(adapterPosition)
                    }
                }
                notesContainer.addView(loadMore)
            }

            folderName.setOnClickListener {
                val newMode =
                    if (expandMode == ExpandMode.COLLAPSED) ExpandMode.FULL else ExpandMode.COLLAPSED
                folderStates[folderId] = newMode
                notifyItemChanged(adapterPosition)
            }

            folderName.setOnLongClickListener {
                val editText = EditText(context)
                editText.setText(folderText)
                editText.imeOptions = EditorInfo.IME_ACTION_DONE

                (folderName.parent as ViewGroup).removeView(folderName)
                (folderName.parent as ViewGroup).addView(editText, 0)

                editText.requestFocus()
                editText.setOnEditorActionListener { v, actionId, _ ->
                    if (actionId == EditorInfo.IME_ACTION_DONE) {
                        val newName = v.text.toString().trim()
                        folderWithNotes.folder.name = newName
                        Log.d("FolderAdapter", "Folder renamed to: $newName")
                        (editText.parent as ViewGroup).removeView(editText)
                        (editText.parent as ViewGroup).addView(folderName, 0)
                        folderName.text = newName
                        true
                    } else false
                }
                true
            }
        }
    }
}
